### DISCUSSION TITLE
Custom Endpoints - Record Hooks and Filtering

### DISCUSSION THREAD

**Discussion by qwacko** [pocketbase/pocketbase#2882] [Q&A] [answered]
Enjoying working with Pocketbase, and I am using it as a framework and it is a pretty great starting point / base. I am adding a few custom endpoints (currently mostly related to bulk actions, but likely to expand over time), however I have run into a couple of items that I may not be completely understanding and was interested in some thoughts / feedback on:

- **Hooks Don't Execute When Records Are Modified In A Custom Endpoint**
 When making a bulk creation endpoint, I was expecting that the `OnRecordBeforeCreateRequest` hook would be called when creating a record, but it doesn't seem to be. Example Below:
```go
e.Router.AddRoute(echo.Route{
			Method: http.MethodPost,
			Path:   "/api/custom/addBulk",
			Middlewares: []echo.MiddlewareFunc{
				apis.ActivityLogger(app),
				adminRoleRequired(app),
			},
			Handler: func(c echo.Context) (err error) {
				u := new(TestBulkData)
				collection, _ := app.Dao().FindCollectionByNameOrId("collectionName")
				if err := c.Bind(u); err != nil {
					log.Print("Data Load Error : ", err)
					return err
				}
				txError := app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
					for _, thisData  := range u.Data {

						record := models.NewRecord(collection)
						form := forms.NewRecordUpsert(app, record)
						form.SetDao(txDao)
						form.LoadData(map[string]any{
							"date":        thisData.Date,
							"description": thisData.Description,
						
						})
						if err := form.Submit(); err != nil {
							log.Print("Error Storing Data : ", err)
							return err
						}
					}
					return nil
				})
				if txError != nil {
					return txError
				}
				return c.String(http.StatusOK, "Complete")
			},
		})
```
Is there possibly something that I am missing here, I wanted the form validation, so I did it that way from the docs, and I assumed that it would call the hooks. The hooks that I have created do execute correctly when I create a record though the normal route. Possibly i need to extract the hook functionality into a function and call that from the custom endpoint as well as the hook so that same logic executes?

- **User Permissions**
When creating a custom hook, I would like the user to only be able to undertake creation / deletion / modification of records that they would normally be able to, but I believe from my testing that this is not the case by default. Is there a way to call functions from a custom endpoint that recognises the api filtering? I have figure how to get the text of the API filter, but not how to then correctly execute it. Possilby there is something that I am missing in the way I call endpoints etc.. . In the above example, if I make it so only admin can create records, the bulk creating endpoint will still allow them to do this.

Possibly for both of these, I just need to create duplicate logic in my custom endpoints to what exists in the pocketbase system, which is OK, just wanted to check if there was any smarter way to achieve this.

**Comment by ganigeorgiev** [comment, marked as answer]
> Hooks Don't Execute When Records Are Modified In A Custom Endpoint ...

`OnRecordBeforeCreateRequest` (and any other `*Request` hook) is triggered only on an API request to the specified action. If you want to attach a general before create logic to a record not depending on the request data you can use the [`OnModelBeforeCreate` hook](https://pocketbase.io/docs/event-hooks/#db-hooks).

> Is there a way to call functions from a custom endpoint that recognises the api filtering? ...

With the upcoming v0.17.0 release there will be a [`Dao.CanAccessRecord()` helper](https://github.com/pocketbase/pocketbase/blob/7bb33d4453453731cd01d6d810105518cd405593/daos/record.go#L498) (_the declaration may slightly change before the release_).

**Reply by qwacko** [reply]
Great thanks for the help / info. Fantastic that the helper will be added, that will be super useful. Actually the link to the planned code would mean that I could probably implement it anyway temporarily, but I think that I can wait for a bit until v0.17.0 comes out.

Appreciate the support.
